Proposal for new interfaces assign strategy

noblepaul opened the pull request.

noblepaul committed:
scratch

noblepaul committed:
javadocs

noblepaul committed:
javadocs

noblepaul committed:
AL header

noblepaul committed:
javadocs

noblepaul committed:
javadocs

sigram commented on solr/solrj/src/java/org/apache/solr/common/assign/ReplicaAssignStrategy.java:
```diff
@@ -0,0 +1,132 @@
+    interface AddOperation extends Operation {
+        String targetNode();
+        String collection();
+        String slice();
```
Let's use "shard" everywhere, there's already too much confusion between slice and shard.

sigram commented on solr/solrj/src/java/org/apache/solr/common/assign/ReplicaAssignStrategy.java:
```diff
@@ -0,0 +1,132 @@
+    interface MoveOperation extends AddOperation {
+
+        String fromNode();
+        String replicaName();
```
What is the replicaName? there's also a lot of confusion around this currently. Is it the core name? or coreNode name? Let's clarify it and use consistently.

noblepaul replied:
coreNodeName

sigram replied:
is coreNodeName global per cluster, or just the collection? should we use core name instead, which is unique?

noblepaul replied:
`coreNodeName` is the global name. That's unique for a given collection

sigram commented on solr/solrj/src/java/org/apache/solr/common/assign/ReplicaAssignStrategy.java:
```diff
@@ -0,0 +1,132 @@
+     */
+    interface MoveOperation extends AddOperation {
+
+        String fromNode();
```
We should have an optional toNode if the requestor knows exactly where it wants to put the replica.

noblepaul replied:
that must go into the `hints`

sigram commented on solr/solrj/src/java/org/apache/solr/common/assign/ReplicaAssignStrategy.java:
```diff
@@ -0,0 +1,132 @@
+        /**Generic method invocation endpoint for v2 APIs
+         *
+         */
+        String endPoint();
```
What's the end point here? /c/collections? Do we have any others? maybe it should be a default method.

noblepaul replied:
yeah, mostly . But there are other sub paths that support collection admin ops as well

sigram commented on solr/solrj/src/java/org/apache/solr/common/assign/ReplicaAssignStrategy.java:
```diff
@@ -0,0 +1,132 @@
+    /**
+     * This operation adds a replica to a given collection/shard
+     */
+    interface AddOperation extends Operation {
```
Why interface here? I would expect a concrete class that implements interface methods.

noblepaul replied:
Make the impl opaque

sigram commented on solr/solrj/src/java/org/apache/solr/common/assign/ReplicaAssignStrategy.java:
```diff
@@ -0,0 +1,132 @@
+        /** The payload. This will be serialized to JSON and will be psosted to SOlr
+         *
+         */
+        MapWriter payload();
```
Maybe this should be an abstract class, or at least provide a default implementation of this method to write out the common payload elements.

`Operation extends MapWriter` would make sense too.

sigram commented on solr/solrj/src/java/org/apache/solr/common/assign/ReplicaAssignStrategy.java:
```diff
@@ -0,0 +1,132 @@
+/**
+ * The implementation class can be stored in clusterprops.json as follows
+ * {
+ *     assign-strategy : {
```
It should be possible to load and use multiple named strategies. I can imagine a common default strategy that uses Policy framework for some collections, and an override in collection config to use another strategy for other collections where Policy doesn't work well.

noblepaul replied:
I would strongly recommend against that. A violation for one strategy may not be same for another

sigram commented on solr/solrj/src/java/org/apache/solr/common/assign/ReplicaAssignStrategy.java:
```diff
@@ -0,0 +1,132 @@
+    /**
+     * get a list of operations that can be performed for the intents.
+     *
+     * There should be a one-to-one mapping between intent and operation
```
Then we need a NO_OP too. We also need to somehow report that the Intent could not be satisfied, maybe an INVALID op?

noblepaul replied:
Yes, I thought of that . Forgot to add

sigram commented on solr/solrj/src/java/org/apache/solr/common/assign/ReplicaAssignStrategy.java:
```diff
@@ -0,0 +1,132 @@
+        /**
+         * Moves a replica
+         */
+        MOVE(CollectionParams.CollectionAction.MOVEREPLICA);
```
DELETEREPLICA too, for scenarios when you dynamically scale down.

noblepaul replied:
Yes.

sigram commented on solr/solrj/src/java/org/apache/solr/common/assign/ReplicaAssignStrategy.java:
```diff
@@ -0,0 +1,132 @@
+    void init(SolrCloudManager cloudManager);
+
+    /**
+     * get a list of operations that can be performed for the intents.
```
"can" or "will"? This is an important difference in semantics, because if you know you will perform these ops then you should make reservations for the positions. "Can" is much weaker, it doesn't lock out the slots for other concurrent computations.

noblepaul replied:
We do not know if the invoker "will" execute these  operations.

sigram replied:
In other words, the assumption that this method is using the current cluster state that was provided to it and itself it does not make any reservations.

(Just clarifying ... I think that's ok, we could provide it a cluster view that already includes other operations in progress).

murblanc commented:
My main concerns with the proposal is that the plugin code now depends on `SolrCloudManager` and `Suggester.Hint`. From these classes/interfaces it then sees `ClusterStateProvider`, `ClusterState` (that you wanted to [remove](https://the-asf.slack.com/archives/CEKUCUNE9/p1592973533487000?thread_ts=1592973533.487000) Noble) etc.

It's going to either be hard or painful to change code (Autoscaling or other related to cluster state etc) when contributions are not living in the same repository or might not be accessible to us at all and can't be refactored at the same time.

murblanc committed:
Create IlanProposalAssigner.java

Rough example of what I mean when I suggest to separate what the new assignment plugins can touch vs existing code.

murblanc commented:
Just committed to this branch what the beginning of what I'm thinking of would look like...

sigram commented on solr/core/src/java/org/apache/solr/cloud/api/collections/IlanProposalAssigner.java:
```diff
@@ -0,0 +1,121 @@
+
+    int getCountReplicas();
+    long getTotalSizeBytes();
+    long getFreeSizeBytes();
```
We need much more information that that here - sysprops, JVM and OS metrics, etc.

murblanc replied:
Agreed.

sigram commented on solr/core/src/java/org/apache/solr/cloud/api/collections/IlanProposalAssigner.java:
```diff
@@ -0,0 +1,121 @@
+
+  interface AssignerCollectionState {
+    String getCollectionName();
+    Iterator<String> getSliceNames();
```
Please, let's not do this again ... :) Let's just use "shard" everywhere, it's already bad as it is.

murblanc replied:
I'm fine with that. Shall we aim to get rid of Slice in general in Solr code?

sigram replied:
That would be great but it's another issue. I already filed Jira for removing ReplicaInfo, we have Replica that serves exactly the same role.

noblepaul replied:
No sure which is better. The shard is a an abused term. Solr code is littered with both

sigram replied:
Whichever we use, lets just use one consistently. Shard is abused? it always refers to a partition of the collection. Adding Slice was a mistake IMHO, but that was done early on when the terminology was still in flux. Nowadays in my experience people commonly use "shard" not "slice".

sigram commented on solr/core/src/java/org/apache/solr/cloud/api/collections/IlanProposalAssigner.java:
```diff
@@ -0,0 +1,121 @@
+    long getFreeSizeBytes();
+  }
+
+  interface AssignerReplicaInfo {
```
Similarly, we need much more info here, too - things like replica state and replica metrics - stuff like physical size, number of docs, recent use, etc etc.

noblepaul replied:
Yes @sigram we actually need to make a comprehensive list of data needed to compute operations

sigram commented on solr/core/src/java/org/apache/solr/cloud/api/collections/IlanProposalAssigner.java:
```diff
@@ -0,0 +1,121 @@
+    String getCollectionName();
+    Iterator<String> getSliceNames();
+
+    Iterator<AssignerReplicaInfo> getSliceReplicasInfo(String sliceName);
```
I understand the desire to hide as much as possible how these are constructed, but we will frequently need things like `size()`, so maybe better to use `Collection` here and in similar places.

murblanc replied:
My motivation to provide Iterator was for cases when it's likely we don't need the whole set/collection (lazy building). Maybe there aren't such cases.

sigram commented on solr/core/src/java/org/apache/solr/cloud/api/collections/IlanProposalAssigner.java:
```diff
@@ -0,0 +1,121 @@
+  interface AssignementRequest {
+  }
+
+  interface AddReplicaAssignmentRequest {
```
These probably can be concrete classes.

murblanc replied:
The requests should not, no reason to expose an implementation to the plugin.
Maybe the decisions could be concrete classes that the plugin would instantiate though.

noblepaul replied:
interfaces are in general better

sigram commented on solr/core/src/java/org/apache/solr/cloud/api/collections/IlanProposalAssigner.java:
```diff
@@ -0,0 +1,121 @@
+    Replica.Type getType();
+  }
+
+  interface AssignmentDecision {
```
Why not pass in the original AssignmentRequest here? this will be needed for logging, tracking and debugging anyway.

murblanc replied:
I was not assuming a one to one correspondance between assignement requests and decisions. The calling code (Solr calling the plugin) knows what it just passed in the call to computeAssignements() when it gets the list of AssignmentDecision back.

sigram replied:
Even if it's not always a 1:1 correspondence, in cases when it is it would be good to know what was the reason for this decision.

noblepaul replied:
Is it possible that a request may not have a corresponding `decision` ?

murblanc replied:
Or the other way around: a request having multiple decisions? Like create collection could result in multiple replica creations. Just making this up as I write, but it doesn't sound impossible.

sigram replied:
I don't see a problem in either case - we should be able to tie any decisions on the output to a particular intent on the input. In 1 -> N case each Decision would simply refer to the same Intent. In 1 -> null we have a NO_OP Decision. Is N -> M even realistic? I don't think so at the moment.

sigram commented on solr/core/src/java/org/apache/solr/cloud/api/collections/IlanProposalAssigner.java:
```diff
@@ -0,0 +1,121 @@
+  }
+
+  interface AddReplicaAssignmentDecision extends AssignmentDecision {
+    String getNodeName();
```
The other properties could come from the AssignmentRequest referenced here.

noblepaul commented:
> My main concern with the proposal is that the plugin code now depends on SolrCloudManager and Suggester.Hint. 

Yes. @murblanc That `Hint` enum will be copied over to the new API . I just reused it for sake of simplicity

I'm +1 for making a lighter `interface` instead of `SolrCloudManager`

noblepaul commented:
@murblanc 

I like your proposal. Why don't we make this formal and make a ticket ?

murblanc commented:
@noblepaul give me a few more days please. I'd like to have a better global view of current Autoscaling to make sure this "stop gap" option (that's how I see my proposal) does go in the right longer term direction.
